### PR TITLE
direct: support forwardingrule target google api bundle

### PR DESCRIPTION
**Pull request by gemmahou**
### Change description

Fixes b/342035994

Support target google api bundle(all-apis/vpc-rc) directly

### Tests you have done

- [x] Run `make ready-pr` to ensure this PR is ready for review.
- [x] Perform necessary E2E testing for changed resources.

### CODE REVIEW

File: pkg/clients/generated/apis/compute/v1beta1/computeforwardingrule_types.go [hunk 1]
```diff
@@ -36,12 +36,10 @@ import (
 )
 
 type ForwardingruleFilterLabels struct {
-	/* Immutable. Name of the metadata label. The length must be between
-	1 and 1024 characters, inclusive. */
+	/* Immutable. Name of the metadata label. The length must be between 1 and 1024 characters, inclusive. */
```
Comment (gemmahou) [inline pkg/clients/generated/apis/compute/v1beta1/computeforwardingrule_types.go:39]: I'm unsure why the format changed in generated type...🥲 I do not think that will break anything, do we want to fix this?

File: pkg/test/constants/constants.go [hunk 2]
```diff
@@ -49,6 +49,8 @@ var TestNameRegexesToSkip = []string{
 	// Disable due to TF bug https://github.com/hashicorp/terraform-provider-google/issues/16255.
 	// We can't specify labels in the create operation, that causes AssertLabelsMatchAndHaveManagedLabel check to fail.
 	".*(regionalforwardingrulepsc).*",
+	// This test only works with direct controller. Re-enable it when we turn on direct ComputeForwardingRule.
+	".*(globalforwardingrulepscgoogleapis).*",
```
Comment (gemmahou) [inline pkg/test/constants/constants.go:53]: Turned on direct controller locally and verified the test passed:

--- PASS: TestCreateNoChangeUpdateDelete (0.13s)
    --- PASS: TestCreateNoChangeUpdateDelete/compute (0.00s)
        --- PASS: TestCreateNoChangeUpdateDelete/compute/basic-globalforwardingrulepscgoogleapis (158.25s)
PASS

File: kubebuilder_2.3.1__.tar.gz
Comment (yuwenma) [inline kubebuilder_2.3.1__.tar.gz:1]: nit: clean up this file.

File: pkg/controller/direct/compute/forwardingrule_controller.go [hunk 3]
```diff
@@ -282,7 +283,13 @@ func (a *forwardingRuleAdapter) Create(ctx context.Context, createOp *directbase
 		return mapCtx.Err()
 	}
 	forwardingRule.Name = direct.LazyPtr(a.id.forwardingRule)
-	forwardingRule.Labels = desired.Labels
+	target := direct.ValueOf(forwardingRule.Target)
+
+	// API restriction: Labels are invalid in Private Service Connect Forwarding Rule.
+	// TF workaround: https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/944
```
Comment (yuwenma) [inline pkg/controller/direct/compute/forwardingrule_controller.go:289]: mind updating the comment? "TF workaround" --> "ConfigConnector workaround on TF-based resource"

File: pkg/test/resourcefixture/testdata/basic/compute/v1beta1/computeforwardingrule/globalcomputeforwardingrule/_http.log [hunk 4]
```diff
@@ -1041,7 +1041,7 @@ X-Xss-Protection: 0
   "insertTime": "2024-04-01T12:34:56.123456Z",
   "kind": "compute#operation",
   "name": "${operationID}",
-  "operationType": "SetTarget",
```
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/compute/v1beta1/computeforwardingrule/globalcomputeforwardingrule/_http.log:1044]: Does the real GCP use `setTarget` on this field?

File: tests/apichecks/testdata/exceptions/acronyms.txt [hunk 5]
```diff
@@ -114,6 +114,7 @@
 [acronyms] crd=computefirewallpolicies.compute.cnrm.cloud.google.com version=v1beta1: field ".status.selfLinkWithId" should be ".status.selfLinkWithID"
 [acronyms] crd=computeforwardingrules.compute.cnrm.cloud.google.com version=v1beta1: field ".spec.sourceIpRanges" should be ".spec.sourceIPRanges"
 [acronyms] crd=computeforwardingrules.compute.cnrm.cloud.google.com version=v1beta1: field ".spec.sourceIpRanges[]" should be ".spec.sourceIPRanges[]"
+[acronyms] crd=computeforwardingrules.compute.cnrm.cloud.google.com version=v1beta1: field ".spec.target.googleApiBundle" should be ".spec.target.googleAPIBundle"
```
Comment (yuwenma) [inline tests/apichecks/testdata/exceptions/acronyms.txt:117]: I was trying to suggest renaming this field to APIBundle. Do you mind explaining why making this an exception? I think the googleAPIBundle is new, right?

File: pkg/test/resourcefixture/testdata/basic/compute/v1beta1/computeforwardingrule/globalcomputeforwardingrule/_http.log (hunk 4 above)
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/compute/v1beta1/computeforwardingrule/globalcomputeforwardingrule/_http.log:1044]: Yes, so I updated the mock service and re-collected the log. This won't affect anything, I just wanted to ensure that the log remains as identical as possible.

**Comment by gemmahou** [comment]
@yuwenma I updated this PR to handle labels better to match realGCP w/TF as much as possible(however, I feel labels for psc forwarding rule is a strange behavior from GCP, we can improve the UX later if there're Cx issues).

Regarding the name of this new field, I checked some doc and found that gcloud used `[target-google-apis-bundle](https://cloud.google.com/sdk/gcloud/reference/compute/forwarding-rules/create#--target-google-apis-bundle)`, so I feel field name `googleAPIBundle` is fine, or I can change to `targetGoogleAPIBundle`. WDYT?

I also partially enabled direct controller when target is set to "google api bundle". @jasonvigil would you mind taking a look?

File: mockgcp/mockcompute/regionalforwardingrulev1.go [hunk 6]
```diff
@@ -82,6 +82,9 @@ func (s *RegionalForwardingRulesV1) Insert(ctx context.Context, req *pb.InsertFo
 	if obj.NetworkTier == nil {
 		obj.NetworkTier = PtrTo("PREMIUM")
 	}
+	if *obj.LoadBalancingScheme == "" {
+		obj.LoadBalancingScheme = nil
+	}
```
Comment (yuwenma) [inline mockgcp/mockcompute/regionalforwardingrulev1.go:88]: can I ask why we need this for the api bundle?

File: pkg/controller/direct/compute/forwardingrule_controller.go [hunk 7]
```diff
@@ -38,13 +40,32 @@ import (
...
-	registry.RegisterModel(krm.ComputeForwardingRuleGVK, NewForwardingRuleModel)
+	rg := &ForwardingRUleReconcileGate{}
+	registry.RegisterModelWithReconcileGate(krm.ComputeForwardingRuleGVK, NewForwardingRuleModel, rg)
 }
 
 func NewForwardingRuleModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
 	return &forwardingRuleModel{config: config}, nil
 }
 
+type ForwardingRUleReconcileGate struct {
```
Comment (yuwenma) [inline pkg/controller/direct/compute/forwardingrule_controller.go:51]: typo: Rule

File: pkg/controller/direct/compute/forwardingrule_controller.go [hunk 8]
```diff
@@ -295,6 +316,15 @@ func (a *forwardingRuleAdapter) Create(ctx context.Context, createOp *directbase
 	forwardingRule.Name = direct.LazyPtr(a.id.forwardingRule)
 	forwardingRule.Labels = desired.Labels
 
+	// API restriction: Cannot set labels during creation(by POST). But it can be set later by PATCH SetLabels.
+	// API error message: Labels are invalid in Private Service Connect Forwarding Rule.
+	// See GH issue for details: https://github.com/hashicorp/terraform-provider-google/issues/16255
```
Comment (yuwenma) [inline pkg/controller/direct/compute/forwardingrule_controller.go:321]: Thanks!

**Review by yuwenma — COMMENTED** [review]
/lgtm
/approve

**Comment by yuwenma** [comment]
> @yuwenma I updated this PR to handle labels better to match realGCP w/TF as much as possible to ensure backward compatibility(however, I feel labels for psc forwarding rule is a strange behavior from GCP, we can improve the UX later if there're Cx issues).
> 
> Regarding the name of this new field, I checked some doc and found that gcloud used [target-google-apis-bundle](https://cloud.google.com/sdk/gcloud/reference/compute/forwarding-rules/create#--target-google-apis-bundle), so I feel we can use `googleAPIsBundle`(instead of `googleAPIBundle`, adding an extra `s`), or `targetGoogleAPIsBundle`. WDYT? (I also commented in the buganizer ticket)
> 

I like the suggestion of using `googleAPIsBundle`, we don't need target because its parent field is `target`

File: mockgcp/mockcompute/regionalforwardingrulev1.go (hunk 6 above)
Comment (gemmahou) [reply mockgcp/mockcompute/regionalforwardingrulev1.go:88]: PSC forwarding rule requires LoadBalancingScheme to be set to `""`. If it's `""`, then it should not appear in the response body. So I add this in order to to match the realGCP log.

**Pull request merged**